### DISCUSSION TITLE
Are the S-Log3 transformations incorrectly applied to legal range data?

### DISCUSSION THREAD

**Discussion by willSmallHD** [colour-science/colour#786] [Q&A] [answered]
Data is transformed into legal range before passing through the log_decoding_SLog3 and log_encoding_SLog3 functions. This is incorrect according to the [Technical Summary for S-Gamut3.Cine/S-Log3 and S-Gamut3/S-Log3 ](https://pro.sony/s3/cms-static-content/uploadfile/06/1237494271406.pdf) document (see FAQ Q4.)

**Comment by nick-shaw** [comment, marked as answer]
That is the purpose of the `out_normalised_code_value` (encoding) and `in_normalised_code_value` (decoding) arguments. You can choose whether to apply a range scaling or not. This is to deal with cases such as recording the S-Log3 SDI output of a Sony camera with a "naive" ProRes Recorder, and then decoding the ProRes as if it were "video", mapping 0% IRE to 0.0 and 100% IRE to 1.0.
```
>>> colour.models.log_encoding_SLog3(0.0, out_normalised_code_value=False)
0.035388127853881277
>>> colour.models.log_encoding_SLog3(0.0, out_normalised_code_value=True)
0.092864125122189639
```
Those two results match the 3.5% IRE and 10-bit CV 95, as given in the table in the Sony paper.
<img width="502" alt="S-log_values" src="https://user-images.githubusercontent.com/12493652/109684101-d9c7b280-7b77-11eb-804b-515f792cf277.png">

Or for all three reflection values given in the table:
```
>>> colour.models.log_encoding_SLog3(np.array([0, 0.18, 0.9]), out_normalised_code_value=False) * 100
array([  3.53881279,  40.63926941,  60.9469472 ])
>>> colour.models.log_encoding_SLog3(np.array([0, 0.18, 0.9]), out_normalised_code_value=True) * 1023
array([  95.        ,  420.        ,  597.89525744])
```
You can do the same with S-Log and S-Log2 to match all the values in the table.

**Reply by thomasmansencal** [reply]
Thanks @nick-shaw! Marking as answered!

**Comment by willSmallHD** [comment]
I agree that that matches the provided table, I will have to go back and understand what I am missing in the documentation. Thank you for clarifying.

**Comment by nick-shaw** [comment]
Believe me, I have spent a lot of time in a camera test rooms with cameras and waveform monitors, working out what is meant by ambiguous documentation!

**Comment by nick-shaw** [comment]
The arguments were originally called `in_legal` and `out_legal`, but were changed because "legal range" and "full range" are potentially ambiguous terms.

**Comment by willSmallHD** [comment]
I understand the arguments. My confusion mostly stems from the difference between slog and slog3. The slog transformation is applied to full range data, whereas the slog3 one is applied to legal.

**Comment by nick-shaw** [comment]
Yes. Sony changed the way they did things with the different S-Log papers. They also changed the handling of "reflectance" vs "linear IRE" just to keep everybody more confused.

I would say S-Log2 and S-Log3 work the same way. They are just described from a different perspective in their respective white papers. Note that if you compare the S-Log(1) and S-Log2 values in the tables in the S-Log2 and S-Log3 white papers, they match.

<img width="451" alt="S-Log2_values" src="https://user-images.githubusercontent.com/12493652/109690870-b48a7280-7b7e-11eb-9619-4aa8339386c4.png">

<img width="502" alt="S-Log_values" src="https://user-images.githubusercontent.com/12493652/109690694-87d65b00-7b7e-11eb-856d-218261c94a4e.png">

**Comment by thomasmansencal** [comment]
@willSmallHD : Welcome!

Given your alias, am I correct thinking you are working for [SmallHD](https://smallhd.com/)?

**Reply by willSmallHD** [reply]
Yes, I am a software developer at SmallHD.

**Reply by thomasmansencal** [reply]
Ah cool! Great gear!
